### PR TITLE
feat: Support non crates.io registries

### PR DESCRIPTION
**Pull request by ncatelli**
## General
This was to start the conversation around working with publishing crates to an [alternate registries](https://doc.rust-lang.org/cargo/reference/registries.html#using-an-alternate-registry). This is useful for teams that make use of semantic-versioning but publish to a private, or non-[crates.io](https://crates.io/) registry.

## Other Changes
- Added `.devcontainers/` to the .gitignore file.
- Corrected a few clippy lints.
  - [redundant borrow](https://github.com/semantic-release-cargo/semantic-release-cargo/compare/master...ncatelli:semantic-release-cargo:feat/support-non-crates.io-registries?expand=1#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R531)
  - [loop indexing](https://github.com/semantic-release-cargo/semantic-release-cargo/compare/master...ncatelli:semantic-release-cargo:feat/support-non-crates.io-registries?expand=1#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R735) warning
- Various documentation fixes.

## TODO
- [x] Handle alternate registry in the verify stage.
- [x] Handle alternate registry in the prepare stage.
- [x] Fix `Release` to not be tied to `crates.io`

## Potential TODO
- [x] Try setting up crates.io as an alternate registry for testing.
- [x] Fix failures in napi feature.
- [x] Generate new napi bindings.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -478,9 +481,8 @@ fn link_is_publishable(link: &PackageLink) -> bool {
 fn package_is_publishable(pkg: &PackageMetadata) -> bool {
     let result = match pkg.publish() {
         guppy::graph::PackagePublish::Unrestricted => true,
-        guppy::graph::PackagePublish::Registries(registries) => {
-            registries.len() == 1 && registries[0] == "crates.io"
-        }
+        guppy::graph::PackagePublish::Registries(registries) if registries.len() == 1 => true,
```
Comment (ncatelli) [inline src/lib.rs:484]: This is a change worth calling out. This removes the hard dependency on `crates.io` being the only valid upstream. I imagine we may want to add some other conditions to this. i.e. Maybe checking for auth tokens. for a particular index?

File: src/lib.rs [hunk 2]
```diff
@@ -478,9 +481,7 @@ fn link_is_publishable(link: &PackageLink) -> bool {
 fn package_is_publishable(pkg: &PackageMetadata) -> bool {
     let result = match pkg.publish() {
         guppy::graph::PackagePublish::Unrestricted => true,
-        guppy::graph::PackagePublish::Registries(registries) => {
-            registries.len() == 1 && registries[0] == "crates.io"
-        }
+        guppy::graph::PackagePublish::Registries(registries) => _registries.len() == 1,
```
Comment (ncatelli) [inline src/lib.rs:484]: This is a change worth calling out. This removes the hard dependency on crates.io being the only valid upstream. I imagine we may want to add some other conditions to this. i.e. Maybe checking for auth tokens. for a particular index?

File: src/lib.rs [hunk 3]
```diff
@@ -478,9 +481,7 @@ fn link_is_publishable(link: &PackageLink) -> bool {
 fn package_is_publishable(pkg: &PackageMetadata) -> bool {
     let result = match pkg.publish() {
         guppy::graph::PackagePublish::Unrestricted => true,
-        guppy::graph::PackagePublish::Registries(registries) => {
-            registries.len() == 1 && registries[0] == "crates.io"
-        }
+        guppy::graph::PackagePublish::Registries(registries) => registries.len() == 1,
```
Comment (ncatelli) [inline src/lib.rs:484]: This is a change worth calling out. This removes the hard dependency on crates.io being the only valid upstream. I imagine we may want to add some other conditions to this. i.e. Maybe checking for auth tokens. for a particular index?

File: src/lib.rs [hunk 4]
```diff
@@ -473,13 +503,15 @@ fn link_is_publishable(link: &PackageLink) -> bool {
...
+/// A package is publishable if either publication is unrestricted or it can be
+/// published to one registry.
+fn package_is_publishable(pkg: &PackageMetadata, registry: Option<&str>) -> bool {
+    let registry_target = registry.unwrap_or("crates.io");
+
     let result = match pkg.publish() {
         guppy::graph::PackagePublish::Unrestricted => true,
         guppy::graph::PackagePublish::Registries(registries) => {
-            registries.len() == 1 && registries[0] == "crates.io"
+            registries.len() == 1 && registries[0] == registry_target
```
Comment (ncatelli) [inline src/lib.rs:514]: To facilitate publishing to private registries, I settled on using the passed argument as check. This defaults to `crates.io` if undefined.

File: src/lib.rs [hunk 5]
```diff
@@ -394,15 +398,41 @@ fn internal_publish(
...
+/// matched against an argument set.
+///
+/// The list of packages will be written to `output`. If `manifest_path` is provided
+/// then it is expected to give the path to the `Cargo.toml` file for the root of the
+/// workspace. If `manifest_path` is `None` then `list_packages` will look for the
+/// root of the workspace in a `Cargo.toml` file in the current directory.
+///
+/// This is a debuging aid and does not directly correspond to a sementic release
+/// step.
+pub fn list_packages_with_arguments(
```
Comment (ncatelli) [inline src/lib.rs:425]: I opted to add an additional public function here with identical behavior to list_packages to support the new registry functionality without introducing a breaking change.

**Comment by EricCrosson** [comment]
Thank you for this contribution, @ncatelli! This code looks great

If you rebase on latest master, you'll pick up a fix for the "Test CLI .*" CI steps. 
Then if we can get the "Build Node-API" tests passing I think we're in good shape to merge

**Comment by ncatelli** [comment]
Thanks @EricCrosson this has been updated. Would there be a way I could trigger workflows just to help iterate faster on your second point about the napi fixes?

**Comment by EricCrosson** [comment]
I just tweaked the repository settings, so hopefully workflow runs will trigger automatically now when your branch is updated.

For even faster iterations, it should be possible to build the native bindings locally. On my x86_64-linux machine, I ran

```shell
npm ci
```

and then this line

https://github.com/semantic-release-cargo/semantic-release-cargo/blob/67f68b1fb20288f62177031b7a3682cd02c24b64/.github/workflows/ci.yml#L122

and I see the same errors that CI is reporting

**Comment by ncatelli** [comment]
Alright @EricCrosson that should have us back to green. Please note I included updates to the `package-lock.json` file. Happy to pull that out if need be. Please allso let me know if you'd like those commits squashed down as well.

**Comment by EricCrosson** [comment]
Looks great to me -- thank you @ncatelli!

**Pull request merged**